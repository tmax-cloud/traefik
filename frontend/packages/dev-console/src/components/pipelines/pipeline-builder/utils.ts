import { history, resourcePathFromModel } from '@console/internal/components/utils';
import { apiVersionForModel } from '@console/internal/module/k8s';
import { ClusterTaskModel, PipelineModel } from '@console/internal/models';
import { Pipeline, PipelineResourceTask, PipelineResourceTaskParam, PipelineTask } from '../../../utils/pipeline-augment';
import { getTaskParameters } from '../resource-utils';
import { TASK_ERROR_STRINGS, TaskErrorType } from './const';
import { PipelineBuilderFormikValues, PipelineBuilderFormValues, TaskErrorMap } from './types';

export const getErrorMessage = (errorTypes: TaskErrorType[], errorMap: TaskErrorMap) => (taskName: string): string => {
  if (!taskName) {
    return TASK_ERROR_STRINGS[TaskErrorType.MISSING_NAME];
  }

  const errorList: TaskErrorType[] | undefined = errorMap?.[taskName];
  if (!errorList) return null;

  const hasRequestedError = errorList.filter(error => errorTypes.includes(error));
  return hasRequestedError.length > 0 ? TASK_ERROR_STRINGS[hasRequestedError[0]] : null;
};

export const taskParamIsRequired = (param: PipelineResourceTaskParam): boolean => {
  return false;
};

export const convertResourceToTask = (resource: PipelineResourceTask, runAfter?: string[]): PipelineTask => {
  return {
    name: resource.metadata.name,
    runAfter,
    taskRef: {
      kind: resource.kind === ClusterTaskModel.kind ? ClusterTaskModel.kind : undefined,
      name: resource.metadata.name,
    },
    params: getTaskParameters(resource).map(param => ({
      name: param.name,
      value: param.default,
      type: param.type
    })),
  };
};

export const getPipelineURL = (namespace: string) => {
  return `/k8s/ns/${namespace}/pipelines`;
};

const removeListRunAfters = (task: PipelineTask, listIds: string[]): PipelineTask => {
  if (task?.runAfter && listIds.length > 0) {
    // Trim out any runAfters pointing at list nodes
    const runAfter = (task.runAfter || []).filter(runAfterName => !listIds.includes(runAfterName));

    return {
      ...task,
      runAfter,
    };
  }

  return task;
};

const removeEmptyDefaultParams = (task: PipelineTask): PipelineTask => {
  if (task.params?.length > 0) {
    // Since we can submit, this param has a default; check for empty values and remove
    // 20210929 Policy changed
    task.params.forEach(param => { if (param.value === undefined) { param.value = (param.type && param.type === 'array') ?  [''] : ''; }; delete param.type; });
    return {
      ...task,
      //params: task.params.filter(param => !!param.value),
    };
  }

  return task;
};

export const convertBuilderFormToPipeline = (formValues: PipelineBuilderFormikValues, namespace: string, existingPipeline?: Pipeline): Pipeline => {
  const { name, resources, params, workspaces, tasks, listTasks, metadata } = formValues;
  const listIds = listTasks.map(listTask => listTask.name);
  const labels = metadata?.labels;

  return {
    ...existingPipeline,
    apiVersion: apiVersionForModel(PipelineModel),
    kind: PipelineModel.kind,
    metadata: {
      ...existingPipeline?.metadata,
      name,
      namespace,
      labels,
    },
    spec: {
      ...existingPipeline?.spec,
      params,
      resources,
      workspaces,
      tasks: tasks.map(task => removeEmptyDefaultParams(removeListRunAfters(task, listIds))),
    },
  };
};

export const convertPipelineToBuilderForm = (pipeline: Pipeline): PipelineBuilderFormValues => {
  if (!pipeline) return null;

  const {
    metadata: { name },
    spec: { params = [], resources = [], workspaces = [], tasks = [] },
  } = pipeline;

  return {
    name,
    params,
    resources,
    workspaces,
    tasks,
    listTasks: [],
  };
};

export const goToYAML = (existingPipeline?: Pipeline, namespace?: string) => {
  history.push(existingPipeline ? `${resourcePathFromModel(PipelineModel, existingPipeline?.metadata?.name, existingPipeline?.metadata?.namespace)}/yaml` : `${getPipelineURL(namespace)}/~new`);
};
