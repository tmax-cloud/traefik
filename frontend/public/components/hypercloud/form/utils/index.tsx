export * from './modal-launcher';
export * from './modal-list';
