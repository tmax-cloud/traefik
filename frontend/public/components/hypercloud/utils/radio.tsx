import * as _ from 'lodash-es';
import * as React from 'react';
import * as classNames from 'classnames';
import { useFormContext } from 'react-hook-form';

export const RadioInput: React.SFC<RadioInputProps> = props => {
  const { register } = props.methods ? props.methods : useFormContext();
  const inputProps: React.InputHTMLAttributes<any> = _.omit(props, ['title', 'subTitle', 'desc', 'children', 'inline']);
  const defaultChecked = props.initvalue === props.item ? true : false;
  const inputElement = (
    <>
      <label className={classNames({ 'radio-inline': props.inline, 'co-disabled': props.disabled })}>
        <input type="radio" ref={register} value={props.item} defaultChecked={defaultChecked} {...inputProps} />
        {props.title} {props.subTitle && <span className="co-no-bold">{props.subTitle}</span>}
      </label>
      {props.desc && <p className="co-m-radio-desc text-muted">{props.desc}</p>}
      {props.children}
    </>
  );

  return props.inline ? inputElement : <div className="radio">{inputElement}</div>;
};

export const RadioGroup: React.SFC<RadioGroupProps> = ({ inline = false, items, id = JSON.stringify(items), name, methods, initValue }) => {
  const radios = items.map(({ desc, title, subTitle, value, disabled }) => <RadioInput name={name} key={value} initvalue={initValue} desc={desc} title={title} subTitle={subTitle} item={value} disabled={disabled} inline={inline} methods={methods} />);
  return <div className={classNames('co-radio-group', { 'co-radio-group--inline': inline })}>{radios}</div>;
};

export type RadioInputProps = {
  item: string;
  desc?: string | JSX.Element;
  subTitle?: string | JSX.Element;
  disabled?: boolean;
  methods?: any;
  inline?: boolean;
  initvalue?: string;
} & React.InputHTMLAttributes<any>;

export type RadioGroupProps = {
  name: string;
  items: ({
    desc?: string | JSX.Element;
    title: string | JSX.Element;
    subTitle?: string | JSX.Element;
    value: string;
    disabled?: boolean;
  } & React.InputHTMLAttributes<any>)[];
  id?: string;
  methods?: any;
  inline?: boolean;
  initValue?: string;
};

RadioInput.displayName = 'RadioInput';
RadioGroup.displayName = 'RadioGroup';
