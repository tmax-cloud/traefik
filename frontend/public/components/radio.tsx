import * as _ from 'lodash-es';
import * as React from 'react';
import * as classNames from 'classnames';

export const RadioInput: React.SFC<RadioInputProps> = props => {
  const inputProps: React.InputHTMLAttributes<any> = _.omit(props, ['title', 'subTitle', 'desc', 'children', 'inline']);
  const { inline, disabled, title, subTitle, desc, children } = props;
  const inputElement = (
    <>
      <label className={classNames({ 'radio-inline': inline, 'co-disabled': disabled })}>
        <input type="radio" {...inputProps} />
        {title} {subTitle && <span className="co-no-bold">{subTitle}</span>}
      </label>
      {desc && <p className="co-m-radio-desc text-muted">{desc}</p>}
      {children}
    </>
  );

  return inline ? inputElement : <div className="radio">{inputElement}</div>;
};

export const RadioGroup: React.SFC<RadioGroupProps> = ({ currentValue, inline = false, items, label, onChange, id = JSON.stringify(items) }) => {
  const radios = items.map(({ desc, title, subTitle, value, disabled }) => <RadioInput key={value} checked={value === currentValue} desc={desc} onChange={onChange} title={title} subTitle={subTitle} value={value} disabled={disabled} inline={inline} />);
  return (
    <div className={classNames('co-radio-group', { 'co-radio-group--inline': inline })}>
      {label ? (
        <>
          <label className="form-label co-radio-group__label" htmlFor={id}>
            {label}
          </label>
          <div className="co-radio-group__controls" id={id}>
            {radios}
          </div>
        </>
      ) : (
        radios
      )}
    </div>
  );
};

export type RadioInputProps = {
  checked: boolean;
  desc?: string | JSX.Element;
  onChange: (v: any) => void;
  subTitle?: string | JSX.Element;
  value: any;
  disabled?: boolean;
  inline?: boolean;
} & React.InputHTMLAttributes<any>;

export type RadioGroupProps = {
  currentValue: any;
  id?: string;
  inline?: boolean;
  items: ({
    desc?: string | JSX.Element;
    title: string | JSX.Element;
    subTitle?: string | JSX.Element;
    value: any;
    disabled?: boolean;
  } & React.InputHTMLAttributes<any>)[];
  label?: string;
  onChange: React.InputHTMLAttributes<any>['onChange'];
};

RadioInput.displayName = 'RadioInput';
RadioGroup.displayName = 'RadioGroup';
