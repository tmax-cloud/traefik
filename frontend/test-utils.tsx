import * as React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from './jest-setup';

// MEMO : redux Provider로 감싼 구조로 @testing-library/react의 render함수를 래핑함.
const render = (ui: React.ReactElement, options?) => {
  const Wrapper = ({ children }: { children?: React.ReactNode }) => {
    return <Provider store={store}>{!!options?.wrapper ? options.wrapper({ children: children }) : children}</Provider>;
  };

  // MEMO : options로 들어온 wrapper는 Wrapper컴포넌트에서 사용하고, 여기서 wrapper속성을 Wrapper로 덮어씌워서 설정해 줌.
  return rtlRender(ui, { ...options, wrapper: Wrapper });
};

export * from '@testing-library/react';
export { render };
