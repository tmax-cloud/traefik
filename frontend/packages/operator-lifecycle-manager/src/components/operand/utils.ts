import * as _ from 'lodash';
import * as Immutable from 'immutable';
import { JSONSchema7 } from 'json-schema';
import { SpecCapability, Descriptor } from '../descriptors/types';
import { modelFor } from '@console/internal/module/k8s';
import { capabilityFieldMap, capabilityWidgetMap } from '../descriptors/spec/spec-descriptor-input';
import { HIDDEN_UI_SCHEMA, REGEXP_K8S_RESOURCE_SUFFIX, REGEXP_SELECT_OPTION, REGEXP_FIELD_DEPENDENCY_PATH_VALUE, SORT_WEIGHT_SCALE_1, SORT_WEIGHT_SCALE_2, SORT_WEIGHT_SCALE_3 } from './const';
import { SchemaType } from '@console/shared/src/components/dynamic-form';
import { getSchemaType } from '@rjsf/core/lib/utils';
import { getSchemaErrors } from '@console/shared/src/components/dynamic-form/utils';
import { isArray } from 'lodash';
import { UiSchema } from '@rjsf/core';

// Transform a path string from a descriptor to a JSON schema path array
export const descriptorPathToUISchemaPath = (path: string): string[] =>
  (_.toPath(path) ?? []).map(subPath => {
    return /^\d+$/.test(subPath) ? 'items' : subPath;
  });

// Determine if a given path is defined on a JSONSchema
export const jsonSchemaHas = (jsonSchema: JSONSchema7, schemaPath: string[]): boolean => {
  const [next, ...rest] = schemaPath;
  const nextSchema = jsonSchema?.[next] ?? jsonSchema?.properties?.[next];
  if (rest.length && !!nextSchema) {
    return jsonSchemaHas(nextSchema, rest);
  }
  return !!nextSchema;
};

// Applies a hidden widget and label configuration to every property of the given schema.
// This is useful for whitelisting only a few schema properties when all properties are not known.
export const hideAllExistingProperties = (schema: JSONSchema7) => {
  return _.reduce(
    schema?.properties,
    (acc, _unused, propertyName) => ({
      ...acc,
      [propertyName]: HIDDEN_UI_SCHEMA,
    }),
    {},
  );
};

// Determine if a schema will produce an empty form field.
export const hasNoFields = (jsonSchema: JSONSchema7 = {}): boolean => {
  // If schema is empty or has unsupported properties, it will not render any fields on the form
  if (getSchemaErrors(jsonSchema).length > 0) {
    return true;
  }

  const type = getSchemaType(jsonSchema) ?? '';
  const handleArray = () => {
    return hasNoFields(jsonSchema.items as JSONSchema7);
  };
  const handleObject = () => {
    return _.every(jsonSchema?.properties, hasNoFields) && !jsonSchema?.additionalProperties;
    //  && hasNoFields(jsonSchema?.additionalProperties as JSONSchema7);
  };

  switch (type) {
    case SchemaType.array:
      return handleArray();
    case SchemaType.object:
      return handleObject();
    // case '':
    //   return true; // type이 없는 parameter에 경우 form editor 안그리는 로직 있던 거 제거
    default:
      return false;
  }
};

const isValidOneofComponent = target => {
  if (_.isArray(target)) {
    return target.every(cur => 'title' in cur);
  } else {
    return 'title' in target;
  }
};

// Map json schema to default ui schema
export const getDefaultUISchema = (jsonSchema: JSONSchema7, jsonSchemaName: string): UiSchema => {
  const type = getSchemaType(jsonSchema ?? {});
  if (hasNoFields(jsonSchema)) {
    return HIDDEN_UI_SCHEMA;
  }
  if (!!jsonSchema?.additionalProperties) {
    if (jsonSchemaName.toLowerCase().indexOf('label') >= 0 || jsonSchemaName.toLowerCase().indexOf('annotation') >= 0 || jsonSchemaName.toLowerCase().indexOf('selector') >= 0) {
      return {
        'ui:field': 'LabelsField',
      };
    }
    return {
      'ui:field': 'AdditionalPropertyField',
    };
  } else if (jsonSchemaName === 'owner' && jsonSchema?.description.indexOf(`Don't deal with this field.`) >= 0) {
    delete jsonSchema?.description;
    delete jsonSchema?.type;
  } else if (jsonSchema?.['x-kubernetes-int-or-string'] || jsonSchema?.['anyOf']) {
    if (
      isArray(jsonSchema.anyOf) &&
      jsonSchema.anyOf.every(cur => {
        if (isArray(cur?.['type'])) {
          return cur['type'][0] === 'string' || cur['type'][0] === 'number' || cur['type'][0] === 'null' || cur['type'][0] === 'integer';
        } else if (cur?.['type']) {
          return cur['type'] === 'string' || cur['type'] === 'number' || cur['type'] === 'null' || cur['type'] === 'integer';
        }
      })
    ) {
      delete jsonSchema?.anyOf;
      jsonSchema.type = 'string';
    }
    delete jsonSchema?.anyOf;
  } else if (jsonSchema?.['oneOf']) {
    if (isValidOneofComponent(jsonSchema.oneOf)) {
      return {
        'ui:field': 'OneOfField', // number or string 일 경우에 컴포넌트
      };
    }
  } else if (jsonSchema?.['allOf']) {
    if (
      isArray(jsonSchema.allOf) &&
      jsonSchema.allOf.every(cur => {
        if (isArray(cur?.['type'])) {
          return cur['type'][0] === 'string' || cur['type'][0] === 'number' || cur['type'][0] === 'null' || cur['type'][0] === 'integer';
        } else if (cur?.['type']) {
          return cur['type'] === 'string' || cur['type'] === 'number' || cur['type'] === 'null' || cur['type'] === 'integer';
        }
      })
    ) {
      delete jsonSchema?.allOf;
      jsonSchema.type = 'string';
    }
    delete jsonSchema?.allOf;
  } else if (jsonSchema?.format === 'data-url' && jsonSchema.type === 'string') {
    return {
      'ui:widget': 'fileUploadWidget',
    };
  } else if (jsonSchemaName === 'provider' && jsonSchema.enum?.length > 0) {
    return {
      'ui:field': 'ProviderDropdownField',
    };
  }

  const handleArray = () => {
    const itemsUISchema = getDefaultUISchema(jsonSchema.items as JSONSchema7, '');
    return !_.isEmpty(itemsUISchema) ? { items: itemsUISchema } : {};
  };

  const handleObject = () => {
    return _.reduce(
      jsonSchema.properties,
      (uiSchemaAccumulator: UiSchema, property: JSONSchema7, name: string) => {
        const propertyUISchema = getDefaultUISchema(property, name);
        return _.isEmpty(propertyUISchema)
          ? uiSchemaAccumulator
          : {
              ...(uiSchemaAccumulator ?? {}),
              [name]: propertyUISchema,
            };
      },
      {},
    );
  };
  switch (type) {
    case SchemaType.array:
      return handleArray();
    case SchemaType.object:
      return handleObject();
    default:
      return {};
  }
};

const k8sResourceCapabilityToUISchema = (capability: SpecCapability): UiSchema => {
  const [, groupVersionKind] = capability.match(REGEXP_K8S_RESOURCE_SUFFIX) ?? [];
  const model = groupVersionKind && modelFor(groupVersionKind);
  if (model) {
    return {
      'ui:widget': 'K8sResourceWidget',
      'ui:options': { model, groupVersionKind },
    };
  }
  return {};
};

const fieldDependencyCapabilityToUISchema = (capability: SpecCapability): UiSchema => {
  const [, path, value] = capability.match(REGEXP_FIELD_DEPENDENCY_PATH_VALUE) ?? [];
  if (!!path && !!value) {
    return { 'ui:dependency': { path: descriptorPathToUISchemaPath(path), value } };
  }
  return {};
};

const selectCapabilitiesToUISchema = (capabilities: SpecCapability[]): UiSchema => {
  const items = capabilities.reduce((optionAccumulator, capability) => {
    const [, option] = capability.match(REGEXP_SELECT_OPTION) ?? [];
    return {
      ...optionAccumulator,
      ...(option && { [option]: option }),
    };
  }, {});

  if (!_.isEmpty(items)) {
    return {
      'ui:field': 'DropdownField',
      'ui:items': items,
    };
  }

  return {};
};

// Given an array of SpecCapabilities, return the appropriate corresponding UISchema
export const capabilitiesToUISchema = (capabilities: SpecCapability[] = []) => {
  if (!capabilities?.length) {
    return {};
  }

  const k8sResourceCapability = _.find(capabilities, capability => capability.startsWith(SpecCapability.k8sResourcePrefix));
  if (k8sResourceCapability) {
    return k8sResourceCapabilityToUISchema(k8sResourceCapability);
  }

  const fieldDependencyCapability = _.find(capabilities, capability => capability.startsWith(SpecCapability.fieldDependency));
  if (fieldDependencyCapability) {
    return fieldDependencyCapabilityToUISchema(fieldDependencyCapability);
  }

  const hasSelectOptions = _.some(capabilities, capability => capability.startsWith(SpecCapability.select));
  if (hasSelectOptions) {
    return selectCapabilitiesToUISchema(capabilities);
  }

  const field = _.reduce(
    capabilities,
    (fieldAccumulator, capability) => {
      return fieldAccumulator ?? capabilityFieldMap.get(capability);
    },
    undefined,
  );

  const widget = _.reduce(
    capabilities,
    (widgetAccumulator, capability) => {
      return widgetAccumulator ?? capabilityWidgetMap.get(capability);
    },
    undefined,
  );

  return {
    ...(field && { 'ui:field': field }),
    ...(widget && { 'ui:widget': widget }),
  };
};

// Given a JSONSchema and associated uiSchema, create the appropriat ui schema order property for the jsonSchema.
// Orders properties according to the following rules:
//  - required properties with an associated ui schema come first,
//  - required properties without an associated ui schema next,
//  - optional fields with an associated ui schema next,
//  - all other properties
export const getJSONSchemaOrder = (jsonSchema, uiSchema) => {
  const type = getSchemaType(jsonSchema ?? {});
  const handleArray = () => {
    const descendantOrder = getJSONSchemaOrder(jsonSchema?.items as JSONSchema7, uiSchema?.items);
    return !_.isEmpty(descendantOrder) ? { items: descendantOrder } : {};
  };

  const handleObject = () => {
    const propertyNames = _.keys(jsonSchema?.properties ?? {});
    if (_.isEmpty(propertyNames)) {
      return {};
    }

    // Map control fields to an array so that  an index can be used to apply a modifier to sort weigths of dependent fields
    const controlProperties = _.reduce(
      uiSchema,
      (controlPropertyAccumulator, { 'ui:dependency': dependency }) => {
        const control = _.last(dependency?.path ?? []);
        return !control ? controlPropertyAccumulator : [...controlPropertyAccumulator, control];
      },
      [],
    );

    /**
     * Give a property name a sort wieght based on whether it has a descriptor (uiSchema has property), is required, or is a control
     * field for a property with a field dependency. A lower weight means higher sort order. Fields are weighted according to the following criteria:
     *  - Required fields with descriptor - 0 to 999
     *  - Required fields without descriptor 1000 to 1999
     *  - Optional fields with descriptor 2000 to 2999
     *  - Control fields that don't fit any above - 3000 to 3999
     *  - All other fields - Infinity
     *
     * Within each of the above criteria, fields are further weighted based on field dependency:
     *   - Fields without dependency - base weight
     *   - Control field - base weight  + (nth control field) * 100
     *   - Dependent field - corresponding control field weight + 10
     *
     * These weight numbers are arbitrary, but spaced far enough apart to leave room for multiple levels of sorting.
     */
    const getSortWeight = (property: string): number => {
      // This property's control field, if it exists
      const control = _.last<string>(uiSchema?.[property]?.['ui:dependency']?.path ?? []);

      // A small offset that is added to the base weight so that control fields get sorted last within
      // their appropriate group
      const controlOffset = (controlProperties.indexOf(property) + 1) * SORT_WEIGHT_SCALE_2;

      // If this property is a dependent, it's weight is based on it's control property
      if (control) {
        return getSortWeight(control) + controlOffset + SORT_WEIGHT_SCALE_1;
      }

      const isRequired = (jsonSchema?.required ?? []).includes(property);
      const hasDescriptor = uiSchema?.[property];

      // Required fields with a desriptor are sorted first (lowest weight).
      if (isRequired && hasDescriptor) {
        return SORT_WEIGHT_SCALE_3 + controlOffset;
      }

      // Fields that are required, but have no descriptors get sorted next
      if (isRequired) {
        return SORT_WEIGHT_SCALE_3 * 2 + controlOffset;
      }

      // Optional fields with descriptors get sorted next
      if (hasDescriptor) {
        return SORT_WEIGHT_SCALE_3 * 3 + controlOffset;
      }

      // Control fields that don't fit into any of the above criteria come next
      if (controlOffset > 0) {
        return SORT_WEIGHT_SCALE_3 * 4 + controlOffset;
      }

      // All other fields are sorted in the order in which they are encountered
      // in the schema
      return Infinity;
    };

    const uiOrder = Immutable.Set(propertyNames)
      .sortBy(getSortWeight)
      .toJS();

    return {
      ...(uiOrder.length > 1 && { 'ui:order': uiOrder }),
      ..._.reduce(
        jsonSchema?.properties ?? {},
        (orderAccumulator, property, propertyName) => {
          const descendantOrder = getJSONSchemaOrder(property, uiSchema?.[propertyName]);
          if (_.isEmpty(descendantOrder)) {
            return orderAccumulator;
          }
          return {
            ...orderAccumulator,
            [propertyName]: descendantOrder,
          };
        },
        {},
      ),
    };
  };

  switch (type) {
    case SchemaType.array:
      return handleArray();
    case SchemaType.object:
      return handleObject();
    default:
      return {};
  }
};

// Map a set of spec descriptors to a ui schema
export const descriptorsToUISchema = (descriptors: Descriptor<SpecCapability>[], jsonSchema: JSONSchema7) => {
  const uiSchemaFromDescriptors = _.reduce(
    descriptors,
    (uiSchemaAccumulator, { path, description, displayName, 'x-descriptors': capabilities = [] }) => {
      const uiSchemaPath = descriptorPathToUISchemaPath(path);
      if (!jsonSchemaHas(jsonSchema, uiSchemaPath)) {
        // eslint-disable-next-line no-console
        console.warn('SpecDescriptor path references a non-existent schema property:', path);
        return uiSchemaAccumulator;
      }
      const isAdvanced = _.includes(capabilities, SpecCapability.advanced);
      const capabilitiesUISchema = capabilitiesToUISchema(_.without(capabilities, SpecCapability.advanced));
      return uiSchemaAccumulator.withMutations(mutable => {
        if (isAdvanced) {
          const advancedPropertyName = _.last(uiSchemaPath);
          const pathToAdvanced = [...uiSchemaPath.slice(0, uiSchemaPath.length - 1), 'ui:advanced'];
          const currentAdvanced = mutable.getIn(pathToAdvanced) ?? Immutable.List();
          mutable.setIn(pathToAdvanced, currentAdvanced.push(advancedPropertyName));
        }

        mutable.setIn(
          uiSchemaPath,
          Immutable.Map({
            ...(description && { 'ui:description': description }),
            ...(displayName && { 'ui:title': displayName }),
            ...capabilitiesUISchema,
          }),
        );
      });
    },
    Immutable.Map(),
  ).toJS();
  // return _.merge(uiSchemaFromDescriptors, getJSONSchemaOrder(jsonSchema, uiSchemaFromDescriptors)); // schema 우선순위 정렬 로직
  return _.merge(jsonSchema, uiSchemaFromDescriptors);
};

// Use jsonSchema, descriptors, and some defaults to generate a uiSchema
export const getUISchema = (jsonSchema, providedAPI) => {
  return _.defaultsDeep(
    {
      metadata: {
        ...hideAllExistingProperties(jsonSchema?.properties?.metadata as JSONSchema7),
        name: {
          'ui:title': 'Name',
          'ui:widget': 'TextWidget',
        },
        labels: {
          'ui:title': 'Labels',
          'ui:field': 'LabelsField',
        },
        'ui:options': {
          label: false,
        },
        'ui:order': ['name', 'labels', '*'],
      },
      spec: {
        ...descriptorsToUISchema(providedAPI?.specDescriptors, jsonSchema?.properties?.spec),
        'ui:options': {
          label: false,
        },
      },
      'ui:order': ['metadata', 'spec', '*'],
    },
    getDefaultUISchema(jsonSchema, ''),
  );
};
