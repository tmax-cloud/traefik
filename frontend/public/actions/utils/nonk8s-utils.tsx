import * as _ from 'lodash-es';
import { CustomMenusMap } from '@console/internal/hypercloud/menu/menu-types';
import { getIngressUrl } from '@console/internal/components/hypercloud/utils/ingress-utils';
import { coFetchJSON } from '../../co-fetch';
import { history } from '@console/internal/components/utils/router';

export const getHelmHost = async () => {
  const mapUrl = (CustomMenusMap as any).Helm.url;
  return mapUrl !== '' ? mapUrl : await getIngressUrl('helm-apiserver');
};

export const helmAPI = '/api/kubernetes/apis/helmapi.tmax.io/v1';

export const getKind = (id: string) => {
  return id.split('~~')[0];
};
const getHelmRepo = (id: string) => {
  return id.split('~~')[1];
};

//get object api url 반환
export const nonK8sObjectUrl = async (id: string, namespace: string, name: string) => {
  const kind = getKind(id);
  let host = '';

  switch (kind) {
    case 'HelmRepository':
      return `${helmAPI}/repos/${name}`;
    case 'HelmRelease':
      host = await getHelmHost();
      return `${host}/helm/v1/namespaces/${namespace}/releases/${name}`;
    case 'HelmChart':
      host = await getHelmHost();
      const helmRepo = getHelmRepo(id);
      return `${host}/helm/v1/charts/${helmRepo}_${name}`;
    default:
      return '';
  }
};

//get object 결과 정리
export const nonK8sObjectResult = (kind: string, response: any) => {
  switch (kind) {
    case 'HelmRepository':
      return response.repoInfo[0];
    case 'HelmRelease':
      return response.release[0];
    case 'HelmChart':
      let entriesvalues = Object.values(_.get(response, 'indexfile.entries'));
      return entriesvalues[0][0];
    default:
      return {};
  }
};

//get list api url 반환
export const nonK8sListUrl = async (id: string, query: any) => {
  const kind = getKind(id);
  let host = '';

  switch (kind) {
    case 'HelmRepository':
      return `${helmAPI}/repos`;
    case 'HelmRelease':
      host = await getHelmHost();
      return query?.ns ? `${host}/helm/v1/namespaces/${query.ns}/releases` : `${host}/helm/v1/releases`;
    case 'HelmChart':
      host = await getHelmHost();
      const helmRepo = getHelmRepo(id);
      return helmRepo ? `${host}/helm/v1/charts?repository=${helmRepo}` : `${host}/helm/v1/charts`;
    default:
      return '';
  }
};
//get list 결과 정리
export const nonK8sListResult = async (id: string, response: any) => {
  const kind = getKind(id);
  switch (kind) {
    case 'HelmRepository':
      if (response.error) {
        return [];
      }
      await (async () => {
        await Promise.all(
          response.repoInfo.map(async repoinfo => {
            const helmHost = await getHelmHost();
            if (helmHost === null) {
              history.push('/ingress-check?ingresslabelvalue=helm-apiserver');
            }
            const response = await coFetchJSON(`${helmHost}/helm/v1/charts?repository=${repoinfo.name}`);
            let tempList = [];
            let entriesvalues = Object.values(_.get(response, 'indexfile.entries'));
            entriesvalues.map(value => {
              tempList.push(value[0]);
            });
            repoinfo.charts = tempList;
          }),
        );
      })();
      return response.repoInfo;
    case 'HelmRelease':
      if (response.error) {
        return [];
      }
      return response.release;
    case 'HelmChart':
      if (response.error) {
        return [];
      }
      let tempList = [];
      let entriesvalues = Object.values(_.get(response, 'indexfile.entries'));
      entriesvalues.map(value => {
        tempList.push(value[0]);
      });
      return tempList;
    default:
      return [];
  }
};
