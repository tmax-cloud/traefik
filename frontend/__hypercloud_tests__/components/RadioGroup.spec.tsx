import * as React from 'react';
import { RadioGroup } from '@console/internal/components/hypercloud/utils/radio';
import { render, act } from '../../test-utils';
import userEvent from '@testing-library/user-event';
import { useForm, FormProvider } from 'react-hook-form';

const mockSubmit = jest.fn(data => {});

const resources = [
  {
    title: 'Cpu',
    value: 'cpu',
  },
  {
    title: 'Gpu',
    value: 'gpu',
  },
  {
    title: 'Memory',
    value: 'memory',
  },
];

const renderRadioGroupForm = () => {
  return render(<RadioGroup name="radio-group" items={resources} inline={false} initValue="gpu" />, {
    wrapper: ({ children }) => {
      const methods = useForm();
      return (
        <FormProvider {...methods}>
          <form
            onSubmit={methods.handleSubmit(data => {
              mockSubmit(data);
            })}
          >
            {children}
            <button type="submit">Submit</button>
          </form>
        </FormProvider>
      );
    },
  });
};

describe('RadioGroup Test', () => {
  beforeEach(() => {
    // Empty
  });

  it('초기 렌더 스냅샷 테스트입니다.', () => {
    const { container } = renderRadioGroupForm();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('items props로 받은 item들을 RadioInput아이템으로 렌더링 되어야 합니다.', () => {
    const { getByDisplayValue } = renderRadioGroupForm();
    getByDisplayValue('cpu');
    getByDisplayValue('gpu');
    getByDisplayValue('memory');
  });

  it('initValue로 설정한 input은 default로 체크 돼있어야 합니다.', () => {
    const { getByDisplayValue } = renderRadioGroupForm();
    const gpuRadio = getByDisplayValue('gpu') as HTMLInputElement;
    expect(gpuRadio.checked).toBeTruthy();
  });

  it('Submit 시 보내지는 value 형식 테스트입니다.', async () => {
    const { getByText } = renderRadioGroupForm();

    // MEMO : memory radio button 선택
    userEvent.click(getByText('Memory'));

    await act(async () => userEvent.click(getByText('Submit')));
    expect(mockSubmit).toHaveBeenCalledTimes(1);
    expect(mockSubmit).toHaveBeenCalledWith({ 'radio-group': 'memory' });
  });
});
