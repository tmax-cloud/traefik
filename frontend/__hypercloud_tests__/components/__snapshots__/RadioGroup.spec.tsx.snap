// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup Test 초기 렌더 스냅샷 테스트입니다. 1`] = `
<form>
  <div
    class="co-radio-group"
  >
    <div
      class="radio"
    >
      <label
        class=""
      >
        <input
          initvalue="gpu"
          item="cpu"
          name="radio-group"
          type="radio"
          value="cpu"
        />
        Cpu
         
      </label>
    </div>
    <div
      class="radio"
    >
      <label
        class=""
      >
        <input
          checked=""
          initvalue="gpu"
          item="gpu"
          name="radio-group"
          type="radio"
          value="gpu"
        />
        Gpu
         
      </label>
    </div>
    <div
      class="radio"
    >
      <label
        class=""
      >
        <input
          initvalue="gpu"
          item="memory"
          name="radio-group"
          type="radio"
          value="memory"
        />
        Memory
         
      </label>
    </div>
  </div>
  <button
    type="submit"
  >
    Submit
  </button>
</form>
`;
